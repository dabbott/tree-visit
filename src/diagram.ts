import { boxDiagram } from './diagram/boxDiagram'
import { directoryDiagram } from './diagram/directoryDiagram'
import { IndexPath } from './indexPath'
import { BaseChildrenOptions } from './options'

export type DiagramType = 'directory' | 'box'

export type DiagramOptions<T> = BaseChildrenOptions<T> & {
  getLabel: (node: T, indexPath: IndexPath) => string
  type?: DiagramType
  flattenSingleChildNodes?: boolean
}

/**
 * Generate a diagram of the tree, as a string.
 */
export function diagram<T>(node: T, options: DiagramOptions<T>): string {
  if (options.type === 'box') {
    return boxDiagram(node, options)
  }

  return directoryDiagram(node, options)
}
