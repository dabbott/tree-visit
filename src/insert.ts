import { IndexPath } from './indexPath'
import { map } from './map'
import { MutationBaseOptions } from './options'

export type InsertOptions<T> = MutationBaseOptions<T> & {
  nodes: T[]
  at: IndexPath
}

/**
 * Insert nodes at a given `IndexPath`.
 */
export function insert<T>(node: T, options: InsertOptions<T>) {
  const { nodes, at } = options

  if (at.length === 0) {
    throw new Error(`Can't insert nodes at the root`)
  }

  const parentIndexPath = at.slice(0, -1)
  const index = at[at.length - 1]

  return map(node, {
    ...options,
    // Avoid calling `getChildren` for every node in the tree.
    // Return [] if we're just going to return the original node anyway.
    getChildren: (node, indexPath) => {
      if (isPrefixPath(indexPath, parentIndexPath)) {
        return options.getChildren(node, indexPath)
      }

      return []
    },
    transform: (node, children: T[], indexPath) => {
      if (isPathEqual(indexPath, parentIndexPath)) {
        return options.create(
          node,
          splice(children, index, 0, ...nodes),
          indexPath
        )
      }

      if (isPrefixPath(indexPath, parentIndexPath)) {
        return options.create(node, children, indexPath)
      }

      return node
    },
  })
}

function splice<T>(
  array: T[],
  start: number,
  deleteCount: number,
  ...items: T[]
) {
  return [
    ...array.slice(0, start),
    ...items,
    ...array.slice(start + deleteCount),
  ]
}

function isPathEqual(a: IndexPath, b: IndexPath) {
  if (a.length !== b.length) {
    return false
  }

  return a.every((segment, index) => segment === b[index])
}

function isPrefixPath(prefix: IndexPath, path: IndexPath) {
  if (prefix.length > path.length) {
    return false
  }

  return prefix.every((segment, index) => segment === path[index])
}
