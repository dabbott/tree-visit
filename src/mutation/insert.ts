import { IndexPath } from '../indexPath'
import {
  MutationBaseChildrenOptions,
  MutationBaseEntriesOptions,
  convertChildrenToEntries,
} from '../options'
import { applyOperations, getInsertionOperations } from './operation'

export type InsertChildrenOptions<T> = MutationBaseChildrenOptions<T> & {
  nodes: T[]
  at: IndexPath
}

export type InsertEntriesOptions<
  T,
  PK extends PropertyKey
> = MutationBaseEntriesOptions<T, PK> & {
  nodes: [PK, T][]
  at: PK[]
}

/**
 * Insert nodes at a given `IndexPath`.
 */
export function insert<T>(node: T, options: InsertChildrenOptions<T>): T
export function insert<T, PK extends PropertyKey>(
  node: T,
  options: InsertEntriesOptions<T, PK>
): T
export function insert<T, PK extends PropertyKey>(
  node: T,
  _options: InsertChildrenOptions<T> | InsertEntriesOptions<T, PK>
): T {
  const options = insertOptionsInterop(_options)

  const { nodes, at } = options

  if (at.length === 0) {
    throw new Error(`Can't insert nodes at the root`)
  }

  const state = getInsertionOperations<T, PK>(at, nodes)

  return applyOperations(node, state, options)
}

function insertOptionsInterop<T, PK extends PropertyKey>(
  options: InsertChildrenOptions<T> | InsertEntriesOptions<T, PK>
): InsertEntriesOptions<T, PK> {
  if ('getEntries' in options) return options

  return {
    ...convertChildrenToEntries<T, PK>(options),
    at: options.at as PK[],
    nodes: options.nodes.map((node, index) => [index as PK, node]),
    create(node, entries, path) {
      const children = entries.map(([, child]) => child)
      return options.create(node, children, path as IndexPath)
    },
  }
}
