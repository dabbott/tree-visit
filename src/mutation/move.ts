import { access } from '../access'
import { IndexPath } from '../indexPath'
import {
  MutationBaseChildrenOptions,
  MutationBaseEntriesOptions,
  convertChildrenToEntries,
} from '../options'
import { ancestorIndexPaths } from './ancestors'
import {
  applyOperations,
  getInsertionOperations,
  getRemovalOperations,
} from './operation'

export type MoveChildrenOptions<T> = MutationBaseChildrenOptions<T> & {
  indexPaths: IndexPath[]
  to: IndexPath
}

export type MoveEntriesOptions<
  T,
  PK extends PropertyKey
> = MutationBaseEntriesOptions<T, PK> & {
  indexPaths: PK[][]
  to: PK[]
}

export function move<T>(node: T, options: MoveChildrenOptions<T>): T
export function move<T, PK extends PropertyKey>(
  node: T,
  options: MoveEntriesOptions<T, PK>
): T
export function move<T, PK extends PropertyKey>(
  node: T,
  _options: MoveChildrenOptions<T> | MoveEntriesOptions<T, PK>
): T {
  const options = moveOptionsInterop<T, PK>(_options)

  if (options.indexPaths.length === 0) return node

  for (const indexPath of options.indexPaths) {
    if (indexPath.length === 0) {
      throw new Error(`Can't move the root node`)
    }
  }

  if (options.to.length === 0) {
    throw new Error(`Can't move nodes to the root`)
  }

  const _ancestorIndexPaths = ancestorIndexPaths(options.indexPaths)

  const nodesToInsert = _ancestorIndexPaths.map(
    (indexPath) =>
      [
        indexPath[indexPath.length - 1],
        access(node, { ...options, path: indexPath }),
      ] as [PK, T]
  )

  const operations = getInsertionOperations(
    options.to,
    nodesToInsert,
    getRemovalOperations<T, PK>(_ancestorIndexPaths)
  )

  return applyOperations(node, operations, options)
}

function moveOptionsInterop<T, PK extends PropertyKey>(
  options: MoveChildrenOptions<T> | MoveEntriesOptions<T, PK>
): MoveEntriesOptions<T, PK> {
  if ('getEntries' in options) return options

  return {
    ...convertChildrenToEntries<T, PK>(options),
    indexPaths: options.indexPaths as PK[][],
    to: options.to as PK[],
    create(node, entries, path) {
      const children = entries.map(([, child]) => child)
      return options.create(node, children, path as IndexPath)
    },
  }
}
