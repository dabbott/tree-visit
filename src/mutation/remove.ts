import { IndexPath } from '../indexPath'
import {
  MutationBaseChildrenOptions,
  MutationBaseEntriesOptions,
  convertChildrenToEntries,
} from '../options'
import { applyOperations, getRemovalOperations } from './operation'

export type RemoveChildrenOptions<T> = MutationBaseChildrenOptions<T> & {
  indexPaths: IndexPath[]
}

export type RemoveEntriesOptions<
  T,
  PK extends PropertyKey
> = MutationBaseEntriesOptions<T, PK> & {
  indexPaths: PK[][]
}

/**
 * Remove nodes at the given `IndexPath`s.
 */
export function remove<T>(node: T, options: RemoveChildrenOptions<T>): T
export function remove<T, PK extends PropertyKey>(
  node: T,
  options: RemoveEntriesOptions<T, PK>
): T
export function remove<T, PK extends PropertyKey>(
  node: T,
  _options: RemoveChildrenOptions<T> | RemoveEntriesOptions<T, PK>
): T {
  const options = removeOptionsInterop(_options)

  if (options.indexPaths.length === 0) return node

  for (const indexPath of options.indexPaths) {
    if (indexPath.length === 0) {
      throw new Error(`Can't remove the root node`)
    }
  }

  const operations = getRemovalOperations<T, PK>(options.indexPaths)

  return applyOperations(node, operations, options)
}

function removeOptionsInterop<T, PK extends PropertyKey>(
  options: RemoveChildrenOptions<T> | RemoveEntriesOptions<T, PK>
): RemoveEntriesOptions<T, PK> {
  if ('getEntries' in options) return options

  return {
    ...convertChildrenToEntries<T, PK>(options),
    indexPaths: options.indexPaths as PK[][],
    create(node, entries, path) {
      const children = entries.map(([, child]) => child)
      return options.create(node, children, path as IndexPath)
    },
  }
}
