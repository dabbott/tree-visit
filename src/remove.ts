import { ancestorIndexPaths } from './ancestors'
import { IndexPath } from './indexPath'
import { map } from './map'
import { MutationBaseOptions } from './options'

export type RemoveOptions<T> = MutationBaseOptions<T> & {
  indexPaths: IndexPath[]
}

enum RemovalState {
  remove,
  replace,
}

function getRemovalState(indexPaths: IndexPath[]) {
  const state = new Map<string, RemovalState>()

  // Mark all parents for replacing
  for (const indexPath of indexPaths) {
    for (let i = indexPath.length - 1; i >= 0; i--) {
      const parentKey = indexPath.slice(0, i).join()

      state.set(parentKey, RemovalState.replace)
    }
  }

  // Mark all nodes for removal
  for (const indexPath of indexPaths) {
    state.set(indexPath.join(), RemovalState.remove)
  }

  return state
}

/**
 * Group indexes to remove by their parent index path.
 */
export function getIndexesToRemove(ancestorIndexPaths: IndexPath[]) {
  const indexesToRemove = new Map<string, number[]>()

  for (const indexPath of ancestorIndexPaths) {
    const parentKey = indexPath.slice(0, -1).join()

    const value = indexesToRemove.get(parentKey) ?? []

    value.push(indexPath[indexPath.length - 1])

    indexesToRemove.set(parentKey, value)
  }

  return indexesToRemove
}

/**
 * Remove nodes at the given `IndexPath`s.
 *
 * This function is used internally by `remove` and `move`.
 */
export function removeInternal<T>(
  node: T,
  options: RemoveOptions<T>,
  ancestorIndexPaths: IndexPath[],
  indexesToRemove: Map<string, number[]>
) {
  const { indexPaths } = options

  if (indexPaths.length === 0) return node

  for (const indexPath of indexPaths) {
    if (indexPath.length === 0) {
      throw new Error(`Can't remove the root node`)
    }
  }

  const state = getRemovalState(ancestorIndexPaths)

  return map(node, {
    ...options,
    // Avoid calling `getChildren` for every node in the tree.
    // Return [] if we're just going to return the original node anyway.
    getChildren: (node, indexPath) => {
      const key = indexPath.join()

      switch (state.get(key)) {
        case RemovalState.replace:
          return options.getChildren(node, indexPath)
        case RemovalState.remove:
        default:
          return []
      }
    },
    transform: (node, children: T[], indexPath) => {
      const key = indexPath.join()
      const indexes = indexesToRemove.get(key)

      if (indexes) {
        return options.create(
          node,
          children.filter((_, index) => !indexes.includes(index)),
          indexPath
        )
      }

      if (state.get(key) === RemovalState.replace) {
        return options.create(node, children, indexPath)
      }

      return node
    },
  })
}

/**
 * Remove nodes at the given `IndexPath`s.
 */
export function remove<T>(node: T, options: RemoveOptions<T>) {
  const _ancestorIndexPaths = ancestorIndexPaths(options.indexPaths)
  const indexesToRemove = getIndexesToRemove(_ancestorIndexPaths)

  return removeInternal(node, options, _ancestorIndexPaths, indexesToRemove)
}
