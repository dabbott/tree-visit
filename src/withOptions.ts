import { BaseOptions } from './options'
import { visit, VisitOptions } from './visit'
import { IndexPath } from './indexPath'
import { access, accessPath } from './access'
import { flat } from './flat'
import {
  find,
  findAll,
  findAllIndexPaths,
  findIndexPath,
  FindOptions,
  FindOptionsTyped,
} from './find'
import { DiagramOptions, diagram } from './diagram'

// Omit any keys from the BaseOptions, since they're always applied
// automatically as part of `withOptions`
type WithoutBase<O> = Omit<O, keyof BaseOptions<unknown>>

export type WithOptions<T> = {
  /**
   * Returns a node by its `IndexPath`.
   *
   * The first node is implicitly included in the `IndexPath` (i.e. no need to pass a `0` first in every `IndexPath`).
   */
  access(node: T, indexPath: IndexPath): T

  /**
   * Returns an array of each node in an `IndexPath`.
   *
   * The first node is implicitly included in the `IndexPath` (i.e. no need to pass a `0` first in every `IndexPath`).
   */
  accessPath(node: T, indexPath: IndexPath): T[]

  /**
   * Generate a diagram of the tree, as a string.
   */
  diagram(node: T, getLabel: DiagramOptions<T>['getLabel']): string
  diagram(node: T, options: WithoutBase<DiagramOptions<T>>): string

  /**
   * Find a node matching a predicate function.
   */
  find(node: T, predicate: FindOptions<T>['predicate']): T | undefined
  find(node: T, options: WithoutBase<FindOptions<T>>): T | undefined
  find<S extends T>(
    node: T,
    predicate: FindOptionsTyped<T, S>['predicate']
  ): S | undefined
  find<S extends T>(
    node: T,
    options: WithoutBase<FindOptionsTyped<T, S>>
  ): S | undefined

  /**
   * Find all nodes matching a predicate function.
   */
  findAll(node: T, predicate: FindOptions<T>['predicate']): T[]
  findAll(node: T, options: WithoutBase<FindOptions<T>>): T[]
  findAll<S extends T>(
    node: T,
    predicate: FindOptionsTyped<T, S>['predicate']
  ): S[]
  findAll<S extends T>(
    node: T,
    options: WithoutBase<FindOptionsTyped<T, S>>
  ): S[]

  /**
   * Find the `IndexPath` of a node matching a predicate function.
   */
  findIndexPath(
    node: T,
    predicate: FindOptions<T>['predicate']
  ): IndexPath | undefined
  findIndexPath(
    node: T,
    options: WithoutBase<FindOptions<T>>
  ): IndexPath | undefined

  /**
   * Find the `IndexPath` of all nodes matching a predicate function.
   */
  findAllIndexPaths(
    node: T,
    predicate: FindOptions<T>['predicate']
  ): IndexPath[]
  findAllIndexPaths(node: T, options: WithoutBase<FindOptions<T>>): IndexPath[]

  /**
   * Returns an array containing the root node and all of its descendants.
   *
   * This is analogous to `Array.prototype.flat` for flattening arrays.
   */
  flat(node: T): T[]

  /**
   * Visit each node using preorder DFS traversal.
   */
  visit(node: T, onEnter: NonNullable<VisitOptions<T>>['onEnter']): void

  /**
   * Visit each node using DFS traversal.
   */
  visit(node: T, options: WithoutBase<VisitOptions<T>>): void
}

/**
 * Return every tree utility function with options partially applied.
 *
 * @param baseOptions
 */
export function withOptions<T>(baseOptions: BaseOptions<T>): WithOptions<T> {
  return {
    access: (node, indexPath) => access(node, indexPath, baseOptions),
    accessPath: (node, indexPath) => accessPath(node, indexPath, baseOptions),
    diagram: (
      node: T,
      getLabelOrOptions:
        | DiagramOptions<T>['getLabel']
        | WithoutBase<DiagramOptions<T>>
    ) =>
      typeof getLabelOrOptions === 'function'
        ? diagram(node, { ...baseOptions, getLabel: getLabelOrOptions })
        : diagram(node, { ...baseOptions, ...getLabelOrOptions }),
    find: (
      node: T,
      predicateOrOptions:
        | FindOptions<T>['predicate']
        | WithoutBase<FindOptions<T>>
    ) =>
      typeof predicateOrOptions === 'function'
        ? find(node, { ...baseOptions, predicate: predicateOrOptions })
        : find(node, { ...baseOptions, ...predicateOrOptions }),
    findAll: (
      node: T,
      predicateOrOptions:
        | FindOptions<T>['predicate']
        | WithoutBase<FindOptions<T>>
    ) =>
      typeof predicateOrOptions === 'function'
        ? findAll(node, { ...baseOptions, predicate: predicateOrOptions })
        : findAll(node, { ...baseOptions, ...predicateOrOptions }),
    findIndexPath: (
      node: T,
      predicateOrOptions:
        | FindOptions<T>['predicate']
        | WithoutBase<FindOptions<T>>
    ) =>
      typeof predicateOrOptions === 'function'
        ? findIndexPath(node, { ...baseOptions, predicate: predicateOrOptions })
        : findIndexPath(node, { ...baseOptions, ...predicateOrOptions }),
    flat: (node: T) => flat(node, baseOptions),
    findAllIndexPaths: (
      node: T,
      predicateOrOptions:
        | FindOptions<T>['predicate']
        | WithoutBase<FindOptions<T>>
    ) =>
      typeof predicateOrOptions === 'function'
        ? findAllIndexPaths(node, {
            ...baseOptions,
            predicate: predicateOrOptions,
          })
        : findAllIndexPaths(node, { ...baseOptions, ...predicateOrOptions }),
    visit: (
      node: T,
      onEnterOrOptions:
        | NonNullable<VisitOptions<T>>['onEnter']
        | WithoutBase<VisitOptions<T>>
    ) =>
      typeof onEnterOrOptions === 'function'
        ? visit(node, { ...baseOptions, onEnter: onEnterOrOptions })
        : visit(node, { ...baseOptions, ...onEnterOrOptions }),
  }
}
