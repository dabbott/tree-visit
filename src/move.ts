import { access } from './access'
import { ancestorIndexPaths } from './ancestors'
import { IndexPath } from './indexPath'
import { insert } from './insert'
import { MutationBaseOptions } from './options'
import { getIndexesToRemove, removeInternal } from './remove'

export type MoveOptions<T> = MutationBaseOptions<T> & {
  indexPaths: IndexPath[]
  to: IndexPath
}

function adjustIndex(
  indexPath: IndexPath,
  indexesToRemove: Map<string, number[]>
) {
  const parentIndexPath = indexPath.slice(0, -1)
  const index = indexPath[indexPath.length - 1]
  const removedIndexes = indexesToRemove.get(parentIndexPath.join()) ?? []
  const adjustedIndex = removedIndexes.reduce(
    (index, removedIndex) => (removedIndex < index ? index - 1 : index),
    index
  )

  return [...parentIndexPath, adjustedIndex]
}

export function move<T>(node: T, options: MoveOptions<T>) {
  if (options.indexPaths.length === 0) return node

  if (options.to.length === 0) {
    throw new Error(`Can't move nodes to the root`)
  }

  const _ancestorIndexPaths = ancestorIndexPaths(options.indexPaths)
  const indexesToRemove = getIndexesToRemove(_ancestorIndexPaths)
  const nodesToInsert = _ancestorIndexPaths.map((indexPath) =>
    access(node, indexPath, options)
  )

  node = removeInternal(node, options, _ancestorIndexPaths, indexesToRemove)

  node = insert(node, {
    ...options,
    at: adjustIndex(options.to, indexesToRemove),
    nodes: nodesToInsert,
  })

  return node
}
