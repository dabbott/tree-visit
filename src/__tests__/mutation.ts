import {
  Node,
  createCountGetChildren,
  createNode,
  example,
  getChildren,
  getLabel,
} from '../__mocks__/node'
import { diagram } from '../diagram'
import { insert } from '../insert'
import { move } from '../move'
import { remove } from '../remove'
import { withOptions } from '../withOptions'

describe('insert', () => {
  it('inserts node at start', () => {
    const result = insert(example, {
      at: [0],
      nodes: [{ name: 'x', indexPath: [] }],
      create: createNode,
      getChildren,
    })

    expect(diagram(result, { getChildren, getLabel })).toMatchSnapshot()
  })

  it('inserts node in middle', () => {
    const result = insert(example, {
      at: [1],
      nodes: [{ name: 'x', indexPath: [] }],
      create: createNode,
      getChildren,
    })

    expect(diagram(result, { getChildren, getLabel })).toMatchSnapshot()
  })

  it('inserts node at end', () => {
    const result = insert(example, {
      at: [2],
      nodes: [{ name: 'x', indexPath: [] }],
      create: createNode,
      getChildren,
    })

    expect(diagram(result, { getChildren, getLabel })).toMatchSnapshot()
  })

  it('inserts nested node', () => {
    const result = insert(example, {
      at: [1, 1],
      nodes: [{ name: 'x', indexPath: [] }],
      create: createNode,
      getChildren,
    })

    expect(diagram(result, { getChildren, getLabel })).toMatchSnapshot()
  })

  it('inserts multiple nodes', () => {
    const { getChildrenWithCount, getCount } = createCountGetChildren()

    const result = insert(example, {
      at: [0, 1],
      nodes: [
        { name: 'x', indexPath: [] },
        { name: 'y', indexPath: [] },
        { name: 'z', indexPath: [] },
      ],
      create: createNode,
      getChildren: getChildrenWithCount,
    })

    expect(diagram(result, { getChildren, getLabel })).toMatchSnapshot()
    expect(getCount()).toEqual(2)
  })
})

describe('remove', () => {
  it('removes node at start', () => {
    const result = remove(example, {
      indexPaths: [[0]],
      create: createNode,
      getChildren,
    })

    expect(diagram(result, { getChildren, getLabel })).toMatchSnapshot()
  })

  it('removes node at end', () => {
    const result = remove(example, {
      indexPaths: [[1]],
      create: createNode,
      getChildren,
    })

    expect(diagram(result, { getChildren, getLabel })).toMatchSnapshot()
  })

  it('removes nested node', () => {
    const { getChildrenWithCount, getCount } = createCountGetChildren()

    const result = remove(example, {
      indexPaths: [[1, 1]],
      create: createNode,
      getChildren: getChildrenWithCount,
    })

    expect(diagram(result, { getChildren, getLabel })).toMatchSnapshot()
    expect(getCount()).toEqual(2)
  })

  it('removes multiple nodes', () => {
    const { getChildrenWithCount, getCount } = createCountGetChildren()

    const result = remove(example, {
      indexPaths: [[1], [0, 1], [0, 0]],
      create: createNode,
      getChildren: getChildrenWithCount,
    })

    expect(diagram(result, { getChildren, getLabel })).toMatchSnapshot()
    expect(getCount()).toEqual(2)
  })

  it('removes multiple nodes with common ancestor', () => {
    const { getChildrenWithCount, getCount } = createCountGetChildren()

    const result = remove(example, {
      indexPaths: [[0, 1], [0]],
      create: createNode,
      getChildren: getChildrenWithCount,
    })

    expect(diagram(result, { getChildren, getLabel })).toMatchSnapshot()
    expect(getCount()).toEqual(1)
  })
})

describe('move', () => {
  it('moves node to the same place', () => {
    const result1 = move(example, {
      indexPaths: [[0]],
      to: [0],
      create: createNode,
      getChildren,
    })

    expect(result1).toEqual(example)

    const result2 = move(example, {
      indexPaths: [[1]],
      to: [1],
      create: createNode,
      getChildren,
    })

    expect(result2).toEqual(example)
  })

  it('moves node from 1 to 0', () => {
    const result = move(example, {
      indexPaths: [[1]],
      to: [0],
      create: createNode,
      getChildren,
    })

    expect(diagram(result, { getChildren, getLabel })).toMatchSnapshot()
  })

  it('moves node from 0 to 1', () => {
    const result = move(example, {
      indexPaths: [[0]],
      to: [1],
      create: createNode,
      getChildren,
    })

    expect(diagram(result, { getChildren, getLabel })).toMatchSnapshot()
  })

  it('moves node from 0.0 to 1', () => {
    const result = move(example, {
      indexPaths: [[0, 0]],
      to: [1],
      create: createNode,
      getChildren,
    })

    expect(diagram(result, { getChildren, getLabel })).toMatchSnapshot()
  })

  it('moves node from 0.0 to 1.0', () => {
    const result = move(example, {
      indexPaths: [[0, 0]],
      to: [1, 0],
      create: createNode,
      getChildren,
    })

    expect(diagram(result, { getChildren, getLabel })).toMatchSnapshot()
  })

  it('moves node using old index', () => {
    const example: Node = {
      name: 'a',
      children: [
        {
          name: 'b',
          indexPath: [0],
        },
        {
          name: 'c',
          indexPath: [1],
        },
        {
          name: 'd',
          indexPath: [2],
        },
        {
          name: 'e',
          indexPath: [3],
        },
        {
          name: 'f',
          indexPath: [4],
        },
      ],
      indexPath: [],
    }

    const result = move(example, {
      indexPaths: [[0], [1]],
      to: [3],
      create: createNode,
      getChildren,
    })

    expect(diagram(result, { getChildren, getLabel })).toMatchSnapshot()
  })
})

describe('partially applied', () => {
  const Tree = withOptions({
    getChildren,
    create: createNode,
  })

  it('inserts node', () => {
    const result = Tree.insert(example, {
      at: [1],
      nodes: [{ name: 'x', indexPath: [] }],
    })

    expect(diagram(result, { getChildren, getLabel })).toMatchSnapshot()
  })

  it('removes node', () => {
    const result = Tree.remove(example, {
      indexPaths: [[1]],
    })

    expect(diagram(result, { getChildren, getLabel })).toMatchSnapshot()
  })

  it('moves node', () => {
    const result = Tree.move(example, {
      indexPaths: [[1, 1]],
      to: [0],
    })

    expect(diagram(result, { getChildren, getLabel })).toMatchSnapshot()
  })
})
