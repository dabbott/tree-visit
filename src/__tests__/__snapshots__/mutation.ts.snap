// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insert inserts multiple nodes 1`] = `
"a
├── b
│   ├── b1
│   ├── x
│   ├── y
│   ├── z
│   └── b2
└── c
    ├── c1
    └── c2"
`;

exports[`insert inserts nested node 1`] = `
"a
├── b
│   ├── b1
│   └── b2
└── c
    ├── c1
    ├── x
    └── c2"
`;

exports[`insert inserts node at end 1`] = `
"a
├── b
│   ├── b1
│   └── b2
├── c
│   ├── c1
│   └── c2
└── x"
`;

exports[`insert inserts node at start 1`] = `
"a
├── x
├── b
│   ├── b1
│   └── b2
└── c
    ├── c1
    └── c2"
`;

exports[`insert inserts node in middle 1`] = `
"a
├── b
│   ├── b1
│   └── b2
├── x
└── c
    ├── c1
    └── c2"
`;

exports[`move moves node from 0 to 1 1`] = `
"a
├── b
│   ├── b1
│   └── b2
└── c
    ├── c1
    └── c2"
`;

exports[`move moves node from 0.0 to 1 1`] = `
"a
├── b
│   └── b2
├── b1
└── c
    ├── c1
    └── c2"
`;

exports[`move moves node from 0.0 to 1.0 1`] = `
"a
├── b
│   └── b2
└── c
    ├── b1
    ├── c1
    └── c2"
`;

exports[`move moves node from 1 to 0 1`] = `
"a
├── c
│   ├── c1
│   └── c2
└── b
    ├── b1
    └── b2"
`;

exports[`move moves node using old index 1`] = `
"a
├── d
├── b
├── c
├── e
└── f"
`;

exports[`partially applied inserts node 1`] = `
"a
├── b
│   ├── b1
│   └── b2
├── x
└── c
    ├── c1
    └── c2"
`;

exports[`partially applied moves node 1`] = `
"a
├── c2
├── b
│   ├── b1
│   └── b2
└── c
    └── c1"
`;

exports[`partially applied removes node 1`] = `
"a
└── b
    ├── b1
    └── b2"
`;

exports[`remove removes multiple nodes 1`] = `
"a
└── b"
`;

exports[`remove removes multiple nodes with common ancestor 1`] = `
"a
└── c
    ├── c1
    └── c2"
`;

exports[`remove removes nested node 1`] = `
"a
├── b
│   ├── b1
│   └── b2
└── c
    └── c1"
`;

exports[`remove removes node at end 1`] = `
"a
└── b
    ├── b1
    └── b2"
`;

exports[`remove removes node at start 1`] = `
"a
└── c
    ├── c1
    └── c2"
`;
